lib/FIR/afforestation.cc

```diff
@@ -1317,84 +1337,150 @@ struct FortranIRLowering {
   // CALL translations ...
   const Value *CreateCalleeValue(
       const parser::ProcedureDesignator &designator) {
-    return static_cast<const Value *>(nullptr);
+    return nullptr;
   }
   CallArguments CreateCallArguments(
       const std::list<parser::ActualArgSpec> &arguments) {
     return CallArguments{};
   }
 
+  Expression VariableToExpression(const parser::Variable &var) {
+    auto maybe{semantics::AnalyzeVariable(semanticsContext_, var)};
+    return {std::move(*maybe)};
+  }
+  Expression DataRefToExpression(const parser::DataRef &dr) {
+    auto maybe{semantics::AnalyzeDataRef(semanticsContext_, dr)};
+    return {std::move(*maybe)};
+  }
+  Expression NameToExpression(const parser::Name &name) {
+    auto maybe{semantics::AnalyzeName(semanticsContext_, name)};
+    return {std::move(*maybe)};
+  }
+
+  void handleIntrinsicAssignmentStmt(const parser::AssignmentStmt &stmt) {
+    // TODO: check if allocation or reallocation should happen, etc.
+    auto *value{builder_->CreateExpr(ExprRef(std::get<parser::Expr>(stmt.t)))};
+    auto *addr{builder_->CreateAddr(
+        VariableToExpression(std::get<parser::Variable>(stmt.t)))};
+    builder_->CreateStore(addr, value);
+  }
+  void handleDefinedAssignmentStmt(const parser::AssignmentStmt &stmt) {
+    CHECK(false && "TODO defined assignment");
+  }
+  void handleAssignmentStmt(const parser::AssignmentStmt &stmt) {
+    // TODO: is this an intrinsic assignment or a defined assignment?
+    if (true) {
+      handleIntrinsicAssignmentStmt(stmt);
+    } else {
+      handleDefinedAssignmentStmt(stmt);
+    }
+  }
+
+  struct AllocOpts {
+    std::optional<Expression *> mold;
+    std::optional<Expression *> source;
+    std::optional<Expression> stat;
+    std::optional<Expression> errmsg;
```

Comment:
Couldn't all four of these be the same type? Either `Expression *` or `std::optional<Expression>`?

Reply:
I did try that but ran into some problems that I can't quite recall now. I'll make a note to revisit.